### ISSUE TITLE
Refine nothings

### ISSUE THREAD

**Issue by mjgpy3** [freckle/maybe-js#14]
I don't really know if this is possible in TypeScript so I'm throwing it here to start discussion.

# Background

In flow we tended to use `?Foo` which ends up being `Foo | null | undefined` in TypeScript. Having two values for nothing (e.g. `null` and `undefined`) is unnecessary and yields a lot of unnecessary checks.

In many places we've started to remove `| undefined` and use `null` to mean "no value".

# Desired behavior

Combinators should not add extra types but should preserve the fidelity of the input types.

E.g. the following should compile
```
let a: number | null = getA()
let b: number | undefined = getB()
let c: number | null | undefined = getC()

const add1 = (v: number) => v+1;

a = mmap(add1, a)
b = mmap(add1, b)
c = mmap(add1, c)
```

**Comment by stackptr** [comment]
This seems to be possible when using [function overloads][0] in an [example Playground][1]:

```
declare function mmap<I extends any, O>(f: (v: I) => O, v?: I): O
declare function mmap<I extends any, O>(f: (v: I) => O, v?: I | null): O | null
```

[0]: https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads
[1]: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMQMBBACgEoAueVZAWwCMQZ4AfW5CCAKFElgIU6bHkLEAQpRp0mLdvDShEWVCGB9w0OEjSZc+IhgDC0znNYc63BUpAq1GnhGLwoMhs0ucbAXnGklM6ujB4WtqjKqurw-kZSFMEY8GBhXgrWEBFRjrEBpok8YHgAzslQwMAAjHlkAG5pLBSxAHzwdQDUVQDcPDwA9P3wAGIwOPTw9FAAnswAtABWJf0lMGD9U7MgAHQYJVQDQ-zaQnqi+PRTAA4APACS8CAAHhggkSVuqNMANPAA8i0yIgaPUaHdmr42n9fnUAPxgjJcCDUf7ZezRYCI7h9QbwADqWAwAAt4Dg6iwIDgKvASlgCKgoBhkHB9poBDphPoxJcoLcHs9Xu9Pj9-oDgfBQfBwa1-jD4VKUX82SddCIDJNrvdHi83sAPlAvr8AUCQQ0FTLoe15Q8rEjFVjeDwoHkeVcyBVqr8oIlGC7ru7KlVfoxEmA-byA56UhQgA

**Comment by stackptr** [comment]
It appears the [existing `mmap` implementation can be overloaded][0]:

```
function mmap<I extends any, O>(f: (v: I) => O, v?: I): O
function mmap<I extends any, O>(f: (v: I) => O, v?: I | null): O | null
function mmap<I extends any, O>(f: (v: I) => O, v?: I | undefined): O | undefined
function mmap<I extends any, O>(f: (v: I) => O, v?: I | null | undefined): O | null | undefined {
  return v === undefined ? undefined : v === null ? null : f(v)
}
```

[0]: https://www.typescriptlang.org/play?ssl=19&ssc=2&pln=14&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMQMBBACgEoAueVZAWwCMQZ4AfW5CCAKFElgIU6bHkLEAQpRp0mLdvDShEWVCGB9w0OEjSZc+IhgDC0znNYc63BUpAq1GnhGLwoMhs0ucbAXnGklM6ujB4WtqjKqurw-kZSFMEY8GBhXgrWEBFRjrEBpok8YHgAzslQwMAAjHlkAG5pLBSxAHzwdQDUVQDcPDwA9P3wAGIwOPTw9FAAnswAtABWJf0lMGD9U7MgAHQYJVQDQ-zaQnqi+PRTAA4APACS8CAAHhggkSVuqNMANPAA8i0yIgaPUaHdmr42n9fnUAPxgjJcCDUf7ZezRYCI7h9YT6MSXKC3B7PV7vT4-f6A4HwUHwcGtf4w+F0lF-Hi486Ta73R4vN7AD5QL6-AFAkENFkM6HtZkPKxI1lY3gcgxcwk8kn8wXCylimkS+mQxkyhEcOwOdSKs2RdGOdlnVUEom80kC8kiqnisEQqFM00+LLWnKWmh-JVoi2YgDePHg8DgGGQMHwdVivn85ox8FhihtkfgNFT6f8mWzAYLSHqiQAvn0oHknWQKtVflBEowG9cm5Uqr9GIkwJ3Cd2WykKEA

**Comment by StevenXL** [comment]
> It appears the [existing `mmap` implementation can be overloaded](https://www.typescriptlang.org/play?ssl=19&ssc=2&pln=14&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMQMBBACgEoAueVZAWwCMQZ4AfW5CCAKFElgIU6bHkLEAQpRp0mLdvDShEWVCGB9w0OEjSZc+IhgDC0znNYc63BUpAq1GnhGLwoMhs0ucbAXnGklM6ujB4WtqjKqurw-kZSFMEY8GBhXgrWEBFRjrEBpok8YHgAzslQwMAAjHlkAG5pLBSxAHzwdQDUVQDcPDwA9P3wAGIwOPTw9FAAnswAtABWJf0lMGD9U7MgAHQYJVQDQ-zaQnqi+PRTAA4APACS8CAAHhggkSVuqNMANPAA8i0yIgaPUaHdmr42n9fnUAPxgjJcCDUf7ZezRYCI7h9YT6MSXKC3B7PV7vT4-f6A4HwUHwcGtf4w+F0lF-Hi486Ta73R4vN7AD5QL6-AFAkENFkM6HtZkPKxI1lY3gcgxcwk8kn8wXCylimkS+mQxkyhEcOwOdSKs2RdGOdlnVUEom80kC8kiqnisEQqFM00+LLWnKWmh-JVoi2YgDePHg8DgGGQMHwdVivn85ox8FhihtkfgNFT6f8mWzAYLSHqiQAvn0oHknWQKtVflBEowG9cm5Uqr9GIkwJ3Cd2WykKEA):
> 
> ```
> function mmap<I extends any, O>(f: (v: I) => O, v?: I): O
> function mmap<I extends any, O>(f: (v: I) => O, v?: I | null): O | null
> function mmap<I extends any, O>(f: (v: I) => O, v?: I | undefined): O | undefined
> function mmap<I extends any, O>(f: (v: I) => O, v?: I | null | undefined): O | null | undefined {
>   return v === undefined ? undefined : v === null ? null : f(v)
> }
> ```

I don't think that overloading is the path I'd take here. Doesn't the code above mean that we are going to end up returning `I | null | undefined` for some `I`, which is the problem we want to get rid of? We are now able to call the function in more ways, but we still end up with the same return type of the `implementation` signature. 

On a tangential point, but while we are here, why do we have `v?` instead of `v`? I don't think I'd like to call `mmap` with only a function from `I to O`?
